cmd/gateway/azure/gateway-azure.go

```diff
@@ -768,19 +790,24 @@ func (a *azureObjects) GetObjectInfo(ctx context.Context, bucket, object string,
 func (a *azureObjects) PutObject(ctx context.Context, bucket, object string, r *minio.PutObjReader, opts minio.ObjectOptions) (objInfo minio.ObjectInfo, err error) {
 	data := r.Reader
 	if data.Size() <= azureBlockSize/2 {
-		blob := a.client.GetContainerReference(bucket).GetBlobReference(object)
-		blob.Metadata, blob.Properties, err = s3MetaToAzureProperties(ctx, opts.UserDefined)
+		blobURL := a.client.NewContainerURL(bucket).NewBlockBlobURL(object)
+		metadata, properties, err := s3MetaToAzureProperties(ctx, opts.UserDefined)
+		if err != nil {
+			return objInfo, azureToObjectError(err, bucket, object)
+		}
+		body, err := ioutil.ReadAll(data)
 		if err != nil {
 			return objInfo, azureToObjectError(err, bucket, object)
 		}
-		if err = blob.CreateBlockBlobFromReader(data, nil); err != nil {
+		_, err = blobURL.Upload(ctx, bytes.NewReader(body), properties, metadata, azblob.BlobAccessConditions{})
```

Comment:
Why not passing 'data' directly here instead of loading it in memory ?

Reply:
`data` is being buffered in memory since `BlockBlobURL.Upload` requires `ReadSeeker` ([see docs](https://godoc.org/github.com/Azure/azure-storage-blob-go/azblob#BlockBlobURL.Upload)) but `r.Reader` only implements `Reader`.

If we want to avoid buffering in memory explicitly, it is also possible to use `azblob.UploadStreamToBlockBlob` ([see docs](https://godoc.org/github.com/Azure/azure-storage-blob-go/azblob#UploadStreamToBlockBlob)) which manages chunking and buffering under the hood. I didn't opt for this choice initially since I aimed to make the code review more straightforward by keeping the new code as close as possible to the old code, but I'm happy to make the change if requested.

Reply:
This does look like a major change which will our gateway use quite a lot of memory for large uploads, we should perhaps not bring regressions here

We would prefer a more memory efficient implementation.

Reply:
@harshavardhana I've switched the implementation to `UploadStreamToBlockBlob` in 983212c which removes all uses of `ioutil.ReadAll` except for the one mentioned in https://github.com/minio/minio/pull/8537#discussion_r349814073. Let me know what you think of this approach.